packages/@cruise-automation/worker-rpc/src/index.js

```diff
@@ -0,0 +1,138 @@
+// instances of web-workers as well as avaiable on 'global' within a worker
+
+export type Channel = {
+  postMessage: (data: any, transfer?: any[]) => void,
+  onmessage?: (ev: MessageEvent) => void,
+};
+
+const RESPONSE = "$$RESPONSE";
+const ERROR = "$$ERROR";
+
+// helper function to create linked channels for testing
+export function createLinkedChannels(): { local: Channel, remote: Channel } {
+  const local: Channel = {
+    postMessage(data: any, transfer?: Array<ArrayBuffer>) {
+      const ev = new MessageEvent("message", { data });
+      // eslint-disable-next-line no-use-before-define
+      if (remote.onmessage) {
+        remote.onmessage(ev); // eslint-disable-line no-use-before-define
+      }
+    },
+  };
+
+  const remote: Channel = {
+    postMessage(data: any, transfer?: Array<ArrayBuffer>) {
+      const ev = new MessageEvent("message", { data });
+      if (local.onmessage) {
+        local.onmessage(ev);
+      }
+    },
+  };
+  return { local, remote };
+}
+
+// This class allows you to hook up bi-directional async calls across web-worker
+// boundaries where a single call to or from a worker can 'wait' on the response.
+// Errors in receivers are propigated back to the caller as a rejection.
+// It also supports returning transferables over the web-worker postMessage api,
+// which was the main shortcomming with the worker-rpc npm module.
+// To attach rpc to an instance of a worker in the main thread:
+//   const rpc = new Rpc(workerInstace);
+// To attach rpc within an a web worker:
+//   const rpc = new Rpc(global);
+// Check out the tests for more examples.
+export default class Rpc {
+  static transferables = "$$TRANSFERABLES";
+  _channel: Channel;
+  _messageId: number = 0;
+  _pendingCallbacks: { [number]: (any) => void } = {};
+  _receivers: Map<string, (any) => any> = new Map();
+
+  constructor(channel: Channel) {
+    this._channel = channel;
+    if (this._channel.onmessage) {
+      throw new Error("channel.onmessage is already set. Can only use one Rpc instance per channel.");
+    }
+    this._channel.onmessage = this._onChannelMessage;
+  }
+
+  _onChannelMessage = (ev: MessageEvent) => {
+    const { id, topic, data } = (ev.data: any);
+    if (topic === RESPONSE) {
+      this._pendingCallbacks[id](ev.data);
+      delete this._pendingCallbacks[id];
+      return;
+    }
+    // invoke the receive handler in a promise so if it throws synchronously we can reject
+    new Promise((resolve, reject) => {
+      const handler = this._receivers.get(topic);
+      if (!handler) {
+        throw new Error(`no receiver registered for ${topic}`);
+      }
+      // This works both when `handler` returns a value or a Promise.
+      resolve(handler(data));
+    })
+      .then((result) => {
+        if (!result) {
+          return this._channel.postMessage({ topic: RESPONSE, id });
+        }
+        const transferables = result[Rpc.transferables];
```

Comment:
I wonder if instead of using this magic string we can just get rid of it and automatically detect transferables by walking the object?

Reply:
Yeah thought about it but there are a few issues w/ that.  Recursively walking the object graph can be expensive.  Extra expensive since structured cloning supports circular references so we'd need a traversal algorithm that supports breaking on circular references.  Since the point of transferables is to be a speed optimization I'm okay w/ having it be an explicit if somewhat more awkward API for performance critical code.  This also more closely parrots the built-in APIs which make you declare what you want to transfer across.  There's also the concern that when you transfer something it becomes detached in the calling code so it could be a surprise if you _want_ to copy a small array buffer or image bitmap and it disappears after you return it from your function and throws next time you access it.